ext/json/php_json_encoder.h

```diff
@@ -26,11 +26,20 @@ struct _php_json_encoder {
 	int depth;
 	int max_depth;
 	php_json_error_code error_code;
+	HashTable recursive;
```

Comment:
Separately from the previous comment, there's the question of whether the json recursion protection should be per-request (in request globals) rather than per call to json_encoder (in request globals, in RINIT/RSHUTDOWN) instead.

My preference is for the former

E.g. a JsonSerializable::jsonSerialize implementation calling `json_encode($this)` would trigger infinite recursion with this PR (but not before this PR) if each call to json_encode had a distinct `recursive` hash table instance - since object property tables don't really change to different pointers in practice if it's necessary for json_encode to work.